File: Code_Implementation/Graphs/Examples/UncertainDestination.java
// # 9370

// (취익)B100 요원, 요란한 옷차림을 한 서커스 예술가 한 쌍이 한 도시의 거리들을 이동하고 있다. 너의 임무는 그들이 어디로 가고 있는지 알아내는 것이다.

// 우리가 알아낸 것은 그들이 s지점에서 출발했다는 것, 그리고 목적지 후보들 중 하나가 그들의 목적지라는 것이다. 그들이 급한 상황이기 때문에 목적지까지

// 우회하지 않고 최단거리로 갈 것이라 확신한다. 이상이다. (취익)

// 어휴! (요란한 옷차림을 했을지도 모를) 듀오가 어디에도 보이지 않는다. 다행히도 당신은 후각이 개만큼 뛰어나다. 이 후각으로 그들이 g와 h 교차로

// 사이에 있는 도로를 지나갔다는 것을 알아냈다. 이 듀오는 대체 어디로 가고 있는 것일까?


// h<->g도로의 가중치를 홀수로 하고 나머지 가중치를 짝수로 만들면 1번의 다익스트라로 풀 수 있다. 덧셈에서 홀수가 한번있으면 무조건 홀수가 되므로 최단거리가
// 홀수가 나오면 h<->g도로를 지난 것이다.
import java.util.*;
class Pair implements Comparable<Pair>{
    int b, w;
    Pair(int b, int w){
        this.b = b;
        this.w = w;
    }

    public int compareTo(Pair p){
        if(w > p.w) return 1;
        else if(w < p.w) return -1;
        else return 0;
    }
}
public class UncertainDestination{
    static ArrayList<ArrayList<Pair>> adj;
    static final int INF = 1000*50000+2;
    public static void Initialize(int n){
        adj = new ArrayList<ArrayList<Pair>>();
        for(int i = 0; i <= n; i++){
            adj.add(new ArrayList<Pair>());
        }
    }

    public static void Add(int a, int b, int w, int g, int h){
        if((a==g&&b==h) || (a==h&&b==g)){
            adj.get(a).add(new Pair(b, w*2-1));
            adj.get(b).add(new Pair(a, w*2-1));
        }
        else {
            adj.get(a).add(new Pair(b, w*2));
            adj.get(b).add(new Pair(a, w*2));
        }
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int cases = sc.nextInt();
        for(int c = 0; c < cases; c++){
            int n = sc.nextInt(), m = sc.nextInt(), t = sc.nextInt();
            int s = sc.nextInt(), g = sc.nextInt(), h = sc.nextInt();
            Initialize(n);
            for(int i = 0; i < m; i++){
                Add(sc.nextInt(), sc.nextInt(), sc.nextInt(),g ,h);
            }
            int[] distance = Dijkstra(s, n);
            ArrayList<Integer> results = new ArrayList<Integer>();
            for(int i = 0; i < t; i++){
                Integer d = sc.nextInt();
                if(distance[d]%2 == 1) results.add(d);
            }

            Collections.sort(results);
            for(Integer i : results){
                System.out.print(i + " ");
            }
            System.out.println();
        }
        sc.close();
    }

    public static int[] Dijkstra(int s, int n){
        int[] distance = new int[n+1];
        boolean[] processed = new boolean[n+1];
        PriorityQueue<Pair> q = new PriorityQueue<>();
        for(int i = 0; i <= n; i++){
            //h<->g도로를 지나지 않으면 전부 짝수가 되어야 하므로 시작점과 연결되는 말든
            //짝수로 남게 한다.
            distance[i] = INF;
        }
        distance[s] = 0;
        q.add(new Pair(s,0));
        while(!q.isEmpty()){
            int a = q.poll().b;
            if(processed[a]) continue;
            processed[a] = true;
            for(Pair p : adj.get(a)){
                if(distance[p.b] > distance[a]+p.w){
                    distance[p.b] = distance[a]+p.w;
                    q.add(new Pair(p.b, distance[p.b]));
                }
            }
        }
        return distance;
    }
}

File: Code_Implementation/Graphs/Examples/UncertainDestination.cs
using System;
using System.Collections.Generic;

public class UD{
    static List<Pair>[] adj;
    static int[] distance;
    static int INF = 50000*1000+2;
    public static void Main(string[] args){
        int tests = int.Parse(Console.ReadLine());
        for(int i = 0; i < tests; i++){
            int[] parts = Array.ConvertAll(Console.ReadLine().Split(' '), s => int.Parse(s));
            int n = parts[0];
            int m = parts[1];
            int t = parts[2];
            parts = Array.ConvertAll(Console.ReadLine().Split(' '), s => int.Parse(s));
            int s = parts[0];
            int g = parts[1];
            int h = parts[2];
            Init(n);
            for(int i = 0; i < m; i++){
                parts = Array.ConvertAll(Console.ReadLine().Split(' '), s => int.Parse(s));
                Add(parts[0],parts[1],parts[2],g,h);
            }
            Dijkstra(s, n);

            List<int> l = new List<int>();
            for(int i = 0; i < t; i++){
                int d = int.Parse(Console.ReadLine());
                if(distance[d]%2==1) l.Add(d);
            }
            l.Sort();
            foreach(var i in l){
                Console.Write(i + " ");
            }
            Console.WriteLine();
        }
    }

    public void Dijkstra(int s, int n){
        for(int i = 1; i <= n; i++){
            distance[i] = INF;
        }
        distance[s] = 0;
        PriorityQueue<Pair> q = new PriorityQueue<Pair>();
        q.Push(new Pair(s,0));
        while(!q.IsEmpty()){
            int a = q.Pop().b;
            foreach(Pair p in adj[a]){
                if(distance[p.b] > distance[a]+p.w){
                    distance[p.b] = distance[a]+p.w;
                    q.Push(new Pair(p.b, distance[p.b]));
                }
            }
        }
    }

    public static void Init(int n){
        distance = new int[n+1];
        adj = new List<Pair>[n+1];
        for(int i = 1; i <= n; i++){
            adj[i] = new List<Pair>();
        }
    }

    public static void Add(int a, int b, int w, int g, int h){
        if((a==h&&b==g) || (a==g&&b==h)){
            adj[a].Add(new Pair(b, w*2-1));
            adj[b].Add(new Pair(a, w*2-1));
        }
        else{
            adj[a].Add(new Pair(b, w*2));
            adj[b].Add(new Pair(a, w*2));
        }
    }
}

class Pair : IComparable<Pair>{
    public int b, w;

    public Pair(int b, int w){
        this.b = b;
        this.w = w;
    }

    public int CompareTo(Pair p){
        return w - p.w;
    }
}

public class PriorityQueue<T>{
    private T[] list;
    public int Count;
    public System.Collections.Generic.Comparer<T> Comparer;

    public PriorityQueue(){
        list = new T[1024];
        Count = 0;
        Comparer = System.Collections.Generic.Comparer<T>.Default;
    }

    public bool IsEmpty(){
        return Count == 0;
    }

    public void Push(T value){
        if(Count == list.Length){
            T[] tmp = new T[list.Length * 2];
            Array.Copy(list, tmp, list.Length);
            list = tmp;
        }
        list[Count++] = value;
        int i = Count - 1;
        while(i > 0 && Comparer.Compare(list[i], list[parent(i)]) < 0){
            T tmp = list[i];
            list[i] = list[parent(i)];
            list[parent(i)] = tmp;
            i = parent(i);
        }
    }

    //가장 말단 노드를 루트 노드에 복사시켜서 그 노드가 다시 맞는 자리로 찾아가도록 한다.
    public T Pop(){
        T ret = list[0];
        list[0] = list[--Count];
        int i = 0;
        while(true){
            int left = (i*2)+1;
            if(left >= Count) break;
            int right = left+1;
            int ind = left;
            //list[0]에 list[--Count]가 대입되었으므로 right == count이면 이미 대입한 값이다.
            if(right < Count && Comparer.Compare(list[left], list[right]) > 0) ind = right;
            if(Comparer.Compare(list[i], list[ind]) >= 0){
                T tmp = list[i];
                list[i] = list[ind];
                list[ind] = tmp;
                i = ind;
            }
            else break;
        }
        return ret;
    }

    private int parent(int i){
        return (i-1)/2;
    }
}